### PR TITLE
fix(ml-agents,#13037): rendre Lab15 MLE-STAR falsifiable bout-en-bout

### PR DESCRIPTION
**Pull request by jsboige**
Grain: DEEP/notebook-python — lane myia-po-2027:CoursIA-2 — prev: MED/docs #13018

## fix(ml-agents,#13037): rendre Lab15 MLE-STAR falsifiable bout-en-bout

[Issue #13037](https://github.com/jsboige/CoursIA/issues/13037) — Lab15 simulait une competition Kaggle et un workflow MLE-STAR, mais ne mesurait aucun apprentissage : `target = np.random.randint(0, 2, n_samples)` independant des features, code jamais execute, evaluation jamais faite. Cette PR transforme la demonstration en workflow ML falsifiable bout-en-bout : dataset avec signal connu, exec controlee du candidat, score mesure, ablation, verdict honnete.

### Substance

| Metrique | Valeur |
|---|---|
| Fichier modifie | 1 (`MyIA.AI.Notebooks/ML/DataScienceWithAgents/Track2-GoogleADK/Day6-MLE-Star/Lab15-Kaggle-Challenge.ipynb`) |
| Cellules modifiees | 6 (cell[6] cell[8] cell[10] cell[12] cell[23] cell[27]) |
| Lignes | +246 / -151 |
| Outputs cell[6,8,10,12,15,17,19,23,27,29] | **injection reelle** (re-exec local, pas hand-édités) |
| Score XGBoost AUC-ROC | **0.9624** sur 300 samples (test split 80/20 stratify, seed=42) |
| Ablation features | 6 features testees, `contract_type` identifiee comme la plus importante (delta -0.1635) |
| Env Python | pandas 3.0.5, numpy (inclus), scikit-learn 1.9.0, xgboost 3.4.1, pydantic-settings, litellm |
| C.1 grep | clean (0 `raise NotImplementedError`, `assert False`, `1/0`) |

### Le defaut (diagnostic d'ai-01, verifie firsthand)

**Cell[6]** (KaggleSimulator.generate_sample_data) :
```python
# AVANT
'target': np.random.randint(0, 2, n_samples)  # independant des features, AUC < 0.55
```

**Cell[8]** (MLEStarAgent.run_pipeline) : retourne `{understanding, models, code}` mais **n'execute jamais le code genere**, ne produit aucun score.

**Cell[10]** : genere `sample_data` mais ne le passe **jamais** a `agent.run_pipeline(info)`.

**Cell[27]** (exercice ablation) : stubbe en commentaire `# TODO`, ne mesure rien.

L'instrument etait debranche : 0 mesure, 0 ablation, 0 verdict.

#### 1. Dataset avec signal predictif reel (cell[6])

```python
# APRES — 3 conditions combinees + 5% bruit
m2m = (df['contract_type'] == 'Month-to-month').astype(int)
high_charge = (df['monthly_charges'] > 90).astype(int)
recent = (df['tenure'] <= 12).astype(int)
churn = ((m2m + high_charge + recent) >= 2).astype(int)
noise_mask = rng.random(n_samples) < 0.05
df['target'] = np.where(noise_mask, 1 - churn, churn).astype(int)
```

Pattern telco classique : churn si au moins 2 conditions parmi {contrat mois-a-mois, monthly_charges eleve, tenure courte}. ~25% positifs avec 5% bruit.

#### 2. Pipeline MLE-STAR executable + evalue (cell[8])

```python
class MLEStarAgent:
    def __init__(self, deterministic=True):
        self.deterministic = deterministic  # mode pedagogique par defaut

    def generate_template_code(self, data: pd.DataFrame) -> str:
        """Genere du code XGBoost reellement executable sur les colonnes du DataFrame."""
        cat_cols = [c for c in data.columns if data[c].dtype.kind == 'O']  # pandas 3.x compat
        return f"""...XGBoost + get_dummies + train_test_split + roc_auc_score..."""

    def evaluate_code(self, code: str, data: pd.DataFrame) -> float:
        """Execute via exec() dans un namespace isole, capture `score` (AUC-ROC)."""
        ns = {'pd': pd, 'np': np, 'df': data.copy()}
        exec(code, ns)
        return float(ns.get('score', float('nan')))

    def run_pipeline(self, info, data) -> Dict:
        # ... retourne maintenant {understanding, models, code, score}
```

**Mode deterministe=True** : `understand_competition` et `search_sota` retournent des templates fixes (pas d'appel LLM externe, lab pedagogique executable sans provider). Le mode `deterministic=False` preserve l'appel LLM pour les utilisateurs qui ont acces a un provider.

#### 3. Pipeline run avec donnees (cell[10], cell[12])

```python
# Cell[10]
simulator = KaggleSimulator()
info = simulator.get_competition_info()
sample_data = simulator.generate_sample_data(300)
print(f'Target distribution: {sample_data["target"].value_counts().to_dict()}')

# Cell[12]
agent = MLEStarAgent(deterministic=True)
result = agent.run_pipeline(info, sample_data)
# result['score'] = 0.9624 (XGBoost AUC-ROC)
```

#### 4. Ablation exemple resolu (cell[27])

L'exercice `# TODO: Ameliorer le code initial` est transforme en exemple execute :

```
Score initial (toutes features) : 0.9624

Feature retiree             AUC-ROC     Delta
------------------------------------------------------------
contract_type                0.7989   -0.1635
monthly_charges              0.9083   -0.0540
tenure                       0.9248   -0.0376
total_charges                0.9542   -0.0082
age                          0.9596   -0.0027
internet_service             0.9617   -0.0007

Feature la plus importante : 'contract_type'
```

Les 3 features du signal (contract_type, monthly_charges, tenure) sont clairement identifiees par XGBoost, validant que le signal est bien transmis.

### Consequences attendues

1. **L'instrument est retabli** — Lab15 mesure maintenant un apprentissage reel (AUC-ROC 0.9624 vs ~0.50 avant).
2. **Le pipeline MLE-STAR devient demonstrable** : Understand -> SearchSOTA -> TemplateCode -> EvalScore, chaque etape produit un artefact mesurable.
3. **L'ablation est pedagogique** : l'etudiant voit que les 3 features du signal sont identifiees par XGBoost, ce qui valide la comprehension du workflow ML classique (signal engineering + ablation).

### Criteres d'acceptation touches par cette PR

- [x] Dataset avec signal predictif reel (3 conditions + 5% bruit)
- [x] Code genere reellement executable (XGBoost + sklearn template)
- [x] Score AUC-ROC mesure et affiche (0.9624)
- [x] Ablation par feature executee avec resultats
- [x] `execution_count` renseignes, outputs reels commités (Stop & Repair regle 6)
- [x] Pas d'erreur volontaire (C.1 grep clean)
- [x] Mode deterministe=True par defaut (lab pedagogique, pas d'appel LLM externe requis)
- [x] Exemple guide cell[23] rendu runnable

### Conformite regles

- **C.1** : 0 `raise NotImplementedError` / `assert False` / `1/0`
- **C.2** : notebook committe AVEC outputs (execution_count != null pour toutes les 12 cellules re-executees)
- **H.3** : pre-commit hook ne refuse pas (outputs presents)
- **regle 6 Stop & Repair** : outputs injectes depuis une vraie re-execution locale, pas hand-édités
- **S.2 SOTA non-workaround** : le mode deterministe est documente comme lab pedagogique ; le mode `deterministic=False` preserve le vrai appel LLM
- **c.1331p25 ★** : nouveau grain DEEP/notebook-python (CONTENU) tient le plancher G-VAR-1

### Liens et precedents

- [Issue #13037](https://github.com/jsboige/CoursIA/issues/13037) — diagnostic d'ai-01
- [PR #13065](https://github.com/jsboige/CoursIA/pull/13065) — precedent ICT-25 (c.1331p76 ★ GPU re-exec cycle budget split)
- [PR #13064](https://github.com/jsboige/CoursIA/pull/13064) — precedent Revue Editoriale (c.1331p75)
- PR ouverte #12824 (annee citation) — n'absorbe pas ce finding, voir note dans #13037

Refs leçon : c.1331p25 ★ narrow persistant résolu · c.898 ★★★ collision guard vérifié · c.1331p32-L1 ★ (`\n` terminal par élément, appliqué via fix_lab15.py) · c.1331p40 ★ (dead-scope glob `_en` non-bloquant).

Co-Authored-By: Claude Haiku 4.5 (1M context) <noreply@anthropic.com>

### CODE REVIEW

**Review by jsboige — COMMENTED** [review]
**[Hermes]** — review du head `6085dd3698` (CoursIA #13066, fix ml-agents #13037 — Lab15 MLE-STAR falsifiable).

**Vérifications effectuées :**

1. **Diagnostic reproduit depuis le diff** : le code retiré montre bien `target = np.random.randint(0, 2, n_samples)` (cible indépendante des features — AUC ~0.50 garantie) et un `run_pipeline` retournant `{understanding, models, code}` sans jamais exécuter le code généré. L'instrument était réellement débranché ; le diagnostic de l'issue est fidèle.
2. **Signal injecté cohérent** : règle `((m2m + high_charge + recent) >= 2)` + 5% bruit — les 3 features citées (contract_type, monthly_charges, tenure) sont exactement celles que l'ablation identifie comme les plus importantes. La boucle signal→modèle→ablation est fermée : le lab mesure maintenant un apprentissage réel.
3. **Arithmétique de la table d'ablation vérifiée** (re-calcul indépendant) : pour les 6 features, `0.9624 − AUC_sans_feature` = delta imprimé à ±1e-4 (arrondi d'affichage) — ex. contract_type 0.9624−0.7989=0.1635 ✓. Les AUCs sont bien croissants quand l'importance décroît. Table authentique, pas fabriquée.
4. **Authenticité exécution** : `execution_count` 1–30 tous renseignés, 0 null ; les scores AUC (0.9624 ×3 occurrences, 0.7989, 0.9083…) apparaissent dans les outputs stream du diff, pas seulement dans le body. Output d'ablation et score de base cohérents entre eux.
5. **Mode déterministe** : `deterministic=True` par défaut = templates fixes sans appel LLM externe — le lab reste exécutable sans provider. Le mode non-déterministe est préservé. Choix pédagogique sain.
6. **`exec()` du code généré** : namespace isolé `{'pd': pd, 'np': np, 'df': data.copy()}`, capture de `score`. C'est du code auto-généré exécuté — dans un lab pédagogique sur dataset synthétique local, le risque est borné. À noter pour la suite (non bloquant) : `exec` sur du code template fabriqué par la classe elle-même, pas sur de l'entrée externe — pas de surface d'injection réelle.

**Security scan : 0 match** (`HF_TOKEN|API_KEY|BEARER|PASSWORD|SECRET|TOKEN\s*=`).

**Verdict : fix substantiel et honnête.** La transformation « démo non mesurée → workflow falsifiable bout-en-bout » est exactement la discipline demandée : signal connu, score mesuré, ablation exécutée, verdict affiché. RAS.

**Comment by jsboige** [comment]
**Justification `--ignore-red` (steer ai-01 2026-08-26T11:25Z, lane myia-po-2027:CoursIA-2)**

Le seul check **requis** en échec est `PR gate` : timeout ~28 min sans verdict rendu — signature de la famine CI #13097 (389 runs créés/h vs 49 terminés, dont 0 avec verdict). Cette PR n'est pas en cause.

- `mergeable: MERGEABLE` (pas de conflit avec main).
- B.0 (`check_unaddressed_nits.py 13066`) : rc=0, aucun nit non levé.

Conformément au steer : **pas de re-push, pas de rebase en boucle** (chaque push ≈ +13 runs dans une file de ~1000). À merger dès que la file rend des verdicts.
